File: examples/dcps/NetworkPartitions/cpp/src/MessageBoard.cpp
/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to 2013 PrismTech
 *   Limited and its licensees. All rights reserved. See file:
 *
 *                     $OSPL_HOME/LICENSE
 *
 *   for full copyright notice and license terms.
 *
 */

/************************************************************************
 * LOGICAL_NAME:    MessageBoard.cpp
 * FUNCTION:        OpenSplice NetworkPartition example code.
 * MODULE:          NetworkPartition example.
 * DATE             june 2007.
 ************************************************************************
 *
 * This file contains the implementation for the 'MessageBoard' executable.
 *
 ***/

#include <iostream>
#include <string.h>
#include "ccpp_dds_dcps.h"
#include "CheckStatus.h"
#include "ccpp_NetworkPartitionsData.h"
#include "example_main.h"
#ifndef _WIN32
#include <unistd.h>
#else
#include "Windows.h"
static void sleep(int secs)
{
    Sleep(secs * 1000);
}
#endif

using namespace DDS;
using namespace NetworkPartitionsData;



#define TERMINATION_MESSAGE -1



int
OSPL_MAIN (
    int argc,
    char *argv[])
{
    /* Generic DDS entities */
    DomainParticipantFactory_var    dpf;
    DomainParticipant_var           participant;
    Topic_var                       chatMessageTopic;
    Subscriber_var                  chatSubscriber;
    DataReader_ptr                  parentReader;

    /* Type-specific DDS entities */
    ChatMessageTypeSupport_var      chatMessageTS;
    ChatMessageDataReader_var       chatAdmin;
    ChatMessageSeq_var              msgSeq = new ChatMessageSeq();
    SampleInfoSeq_var               infoSeq = new SampleInfoSeq();

    /* QosPolicy holders */
    TopicQos                        reliable_topic_qos;
    SubscriberQos                   sub_qos;
    DDS::StringSeq                  parameterList;

    /* DDS Identifiers */
    DomainId_t                      domain = 0;
    ReturnCode_t                    status;

    /* Others */
    bool                            terminated = false;
    const char *                    partitionName = "ChatRoom1";
    char  *                         chatMessageTypeName = NULL;

    /* Options: MessageBoard [ownID] */
    /* Messages having owner ownID will be ignored */
    parameterList.length(1);

    if (argc > 1) {
        parameterList[0] = string_dup(argv[1]);
    }
    else
    {
        parameterList[0] = "0";
    }

    /* Create a DomainParticipantFactory and a DomainParticipant (using Default QoS settings. */
    dpf = DomainParticipantFactory::get_instance();
    checkHandle(dpf.in(), "DDS::DomainParticipantFactory::get_instance");
    participant = dpf->create_participant (
        domain,
        PARTICIPANT_QOS_DEFAULT,
        NULL,
        STATUS_MASK_NONE);
    checkHandle(participant, "DDS::DomainParticipantFactory::create_participant");

    /* Register the required datatype for ChatMessage. */
    chatMessageTS = new ChatMessageTypeSupport();
    checkHandle(chatMessageTS.in(), "new ChatMessageTypeSupport");
    chatMessageTypeName = chatMessageTS->get_type_name();
    status = chatMessageTS->register_type(
        participant.in(),
        chatMessageTypeName);
    checkStatus(status, "NetworkPartitionsData::ChatMessageTypeSupport::register_type");

    /* Set the ReliabilityQosPolicy to RELIABLE. */
    status = participant->get_default_topic_qos(reliable_topic_qos);
    checkStatus(status, "DDS::DomainParticipant::get_default_topic_qos");
    reliable_topic_qos.reliability.kind = DDS::RELIABLE_RELIABILITY_QOS;

    /* Make the tailored QoS the new default. */
    status = participant->set_default_topic_qos(reliable_topic_qos);
    checkStatus(status, "DDS::DomainParticipant::set_default_topic_qos");

    /* Use the changed policy when defining the ChatMessage topic */
    chatMessageTopic = participant->create_topic(
        "Chat_ChatMessage",
        chatMessageTypeName,
        reliable_topic_qos,
        NULL,
        STATUS_MASK_NONE);
    checkHandle(chatMessageTopic.in(), "DDS::DomainParticipant::create_topic (ChatMessage)");

    /* Adapt the default SubscriberQos to read from the "ChatRoom1" Partition. */
    status = participant->get_default_subscriber_qos (sub_qos);
    checkStatus(status, "DDS::DomainParticipant::get_default_subscriber_qos");
    sub_qos.partition.name.length(1);
    sub_qos.partition.name[0] = partitionName;

    /* Create a Subscriber for the MessageBoard application. */
    chatSubscriber = participant->create_subscriber(sub_qos, NULL, STATUS_MASK_NONE);
    checkHandle(chatSubscriber.in(), "DDS::DomainParticipant::create_subscriber");

    /* Create a DataReader for the NamedMessage Topic (using the appropriate QoS). */
    parentReader = chatSubscriber->create_datareader(
        chatMessageTopic.in(),
        DATAREADER_QOS_USE_TOPIC_QOS,
        NULL,
        STATUS_MASK_NONE);
    checkHandle(parentReader, "DDS::Subscriber::create_datareader");

    /* Narrow the abstract parent into its typed representative. */
    chatAdmin = ChatMessageDataReader::_narrow(parentReader);
    checkHandle(chatAdmin.in(), "NetworkPartitionsData::ChatMessageDataReader::_narrow");

    /* Print a message that the MessageBoard has opened. */
    cout << "MessageBoard has opened: send a ChatMessage with userID = -1 to close it...." << endl << endl;

    while (!terminated) {
        /* Note: using read does not remove the samples from
           unregistered instances from the DataReader. This means
           that the DataRase would use more and more resources.
           That's why we use take here instead. */

        status = chatAdmin->take(
            msgSeq,
            infoSeq,
            LENGTH_UNLIMITED,
            ANY_SAMPLE_STATE,
            ANY_VIEW_STATE,
            ALIVE_INSTANCE_STATE );
        checkStatus(status, "NetworkPartitionsData::ChatMessageDataReader::take");

        for (ULong i = 0; i < msgSeq->length(); i++) {
            ChatMessage *msg = &(msgSeq[i]);
            if (msg->userID == TERMINATION_MESSAGE) {
                cout << "Termination message received: exiting..." << endl;
                terminated = true;
            } else {
                cout << msg->userID << ": " << msg->content << endl;
            }
        }

        status = chatAdmin->return_loan(msgSeq, infoSeq);
        checkStatus(status, "NetworkPartitionsData::ChatMessageDataReader::return_loan");

        /* Sleep for some amount of time, so as not to consume too many CPU cycles. */
        sleep(1);
    }

    /* Remove the DataReader */
    status = chatSubscriber->delete_datareader(chatAdmin.in());
    checkStatus(status, "DDS::Subscriber::delete_datareader");

    /* Remove the Subscriber. */
    status = participant->delete_subscriber(chatSubscriber.in());
    checkStatus(status, "DDS::DomainParticipant::delete_subscriber");

    /* Remove the Topic. */
    status = participant->delete_topic(chatMessageTopic.in());
    checkStatus(status, "DDS::DomainParticipant::delete_topic (chatMessageTopic)");

    /* De-allocate the type-names. */
    string_free(chatMessageTypeName);

    /* Remove the DomainParticipant. */
    status = dpf->delete_participant(participant.in());
    checkStatus(status, "DDS::DomainParticipantFactory::delete_participant");

    return 0;
}

File: examples/dcps/Tutorial/java5/src/chatroom/MessageBoard.java
/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to TO_YEAR PrismTech
 *   Limited, its affiliated companies and licensors. All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

package chatroom;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;

import org.omg.dds.core.DDSException;
import org.omg.dds.core.ServiceEnvironment;
import org.omg.dds.core.policy.Durability;
import org.omg.dds.core.policy.History;
import org.omg.dds.core.policy.Partition;
import org.omg.dds.core.policy.PolicyFactory;
import org.omg.dds.core.policy.Reliability;
import org.omg.dds.core.status.Status;
import org.omg.dds.domain.DomainParticipant;
import org.omg.dds.domain.DomainParticipantFactory;
import org.omg.dds.pub.DataWriter;
import org.omg.dds.pub.DataWriterQos;
import org.omg.dds.pub.Publisher;
import org.omg.dds.pub.PublisherQos;
import org.omg.dds.sub.DataReader;
import org.omg.dds.sub.DataReaderListener;
import org.omg.dds.sub.DataReaderQos;
import org.omg.dds.sub.Sample;
import org.omg.dds.sub.Subscriber;
import org.omg.dds.sub.SubscriberQos;
import org.omg.dds.topic.ContentFilteredTopic;
import org.omg.dds.topic.Topic;
import org.omg.dds.topic.TopicQos;

import Chat.ChatMessage;
import Chat.NameService;
import Chat.NamedMessage;

public class MessageBoard {

    private static ServiceEnvironment env = null;
    private static DomainParticipantFactory dpf = null;
    private static DomainParticipant participant = null;
    private static PolicyFactory policyFactory = null;
    private static TopicQos reliableTopicQos = null;
    private static Topic<ChatMessage> chatMessageTopic = null;
    private static Topic<NameService> nameServiceTopic = null;
    private static Topic<NamedMessage> namedMessageTopic = null;
    private static Subscriber subscriber = null;
    private static Publisher publisher = null;
    private static DataReader<ChatMessage> chatMessageDataReader = null;
    private static DataReader<NameService> nameServiceDataReader = null;
    private static DataWriter<NamedMessage> namedMessageDataWriter = null;
    private static ContentFilteredTopic<ChatMessage> cfm =null;
    private static boolean terminated = false;
    public static final int TERMINATION_MESSAGE = -1;
    private static String[] parameterList = new String[1];

    public static void create_simulated_multitopic() {
        /*
         * A Topic is created for the Chat_ChatMessage sample type on the
         * domain participant.
         */
        Collection<Class<? extends Status>> status = new HashSet<Class<? extends Status>>();

        /*
         * A TopicQos is created with Durability set to Transient to
         * ensure that if a subscriber joins after the sample is written then DDS
         * will still retain the sample for it.
         */
        Durability durability = policyFactory.Durability().withTransient();
        TopicQos transientTopicQos = participant.getDefaultTopicQos().withPolicy(durability);

        /*
         * A Topic is created for the Chat_NameService sample type on the
         * domain participant.
         */
        nameServiceTopic = participant.createTopic("Chat_NameService", NameService.class, transientTopicQos, null, status);

        /*
         * A Topic is created for the Chat_NamedMessage sample type on the
         * domain participant.
         */
        namedMessageTopic = participant.createTopic("Chat_NamedMessage", NamedMessage.class, reliableTopicQos, null, status);

        /* A content-filtered topic is created, filtering out our own ID. */
        List<String> p = new ArrayList<String>();
        p.add(parameterList[0]);
        String filter = "userID <> %0";
        cfm = participant.createContentFilteredTopic("Chat_FilteredMessage", chatMessageTopic, filter, p);

        /* Create a KEEP_ALL datareader for the chat-messages */
        DataReaderQos drQos = subscriber.copyFromTopicQos(subscriber.getDefaultDataReaderQos(), reliableTopicQos);
        History history = policyFactory.History().withKeepAll();
        drQos = drQos.withPolicy(history);
        chatMessageDataReader = subscriber.createDataReader(cfm, drQos);

        /* Create a TRANSIENT datareader for the name-service topic */
        drQos = subscriber.copyFromTopicQos(subscriber.getDefaultDataReaderQos(), transientTopicQos);
        nameServiceDataReader = subscriber.createDataReader(nameServiceTopic, drQos);

        /* Create a publisher and datawriter for the simulated multitopic */
        Partition chatroom = policyFactory.Partition().withName("ChatRoom");
        PublisherQos chatPartitionQos = participant.getDefaultPublisherQos().withPolicy(chatroom);
        publisher = participant.createPublisher(chatPartitionQos);
        DataWriterQos dwQos = publisher.copyFromTopicQos(publisher.getDefaultDataWriterQos(), reliableTopicQos);
        namedMessageDataWriter = publisher.createDataWriter(namedMessageTopic,dwQos);

        /* Create and set the listener for the ChatMessage */
        DataReaderListener<ChatMessage> drl = new DataReaderListenerClass(chatMessageDataReader, nameServiceDataReader, namedMessageDataWriter);
        chatMessageDataReader.setListener(drl);
    }

    public static void main(String[] args) {
        /* Options: MessageBoard [ownID] */
        /* Messages having owner ownID will be ignored */
        if (args.length >0) {
            parameterList[0] = args[0];
        }
        else {
            parameterList[0] = new String("0");
        }

        /* The service environment is created for the application. */
        System.setProperty(ServiceEnvironment.IMPLEMENTATION_CLASS_NAME_PROPERTY,
                "org.opensplice.dds.core.OsplServiceEnvironment");

        env = ServiceEnvironment.createInstance(MessageBoard.class.getClassLoader());

        /* A DomainParticipant is created for the default domain. */
        dpf = DomainParticipantFactory.getInstance(env);

        try {
            participant = dpf.createParticipant();

            /*
             * A TopicQos is created with Reliability set to Reliable to
             * guarantee delivery.
             */
            policyFactory = env.getSPI().getPolicyFactory();
            Reliability reliable = policyFactory.Reliability().withReliable();
            reliableTopicQos = participant.getDefaultTopicQos().withPolicy(reliable);
            /* Set the Reliable TopicQos as the new default */
            participant.setDefaultTopicQos(reliableTopicQos);

            /*
             * A Topic is created for the Chat_ChatMessage sample type on the
             * domain participant.
             */
            Collection<Class<? extends Status>> status = new HashSet<Class<? extends Status>>();

            chatMessageTopic = participant.createTopic("Chat_ChatMessage", ChatMessage.class, reliableTopicQos, null, status);

            /*
             * Create a Subscriber and a NamedMessageDataReader whose
             * QoS Settings comply with that on the topic.
             */
            Partition chatroom = policyFactory.Partition().withName("ChatRoom");
            SubscriberQos chatPartitionQos = participant.getDefaultSubscriberQos().withPolicy(chatroom);
            subscriber = participant.createSubscriber(chatPartitionQos);
            create_simulated_multitopic();

            DataReaderQos drQos = subscriber.copyFromTopicQos(subscriber.getDefaultDataReaderQos(), reliableTopicQos);
            History history = policyFactory.History().withKeepAll();
            drQos = drQos.withPolicy(history);
            DataReader<NamedMessage> namedMessageDataReader = subscriber.createDataReader(namedMessageTopic,drQos);

            System.out.println(
                    "MessageBoard has opened: send a ChatMessage " +
                    "with userID = -1 to close it....\n");

            while (!terminated) {
                /* Note: using read does not remove the samples from unregistered
                 * instances from the DataReader. This means that the DataReader
                 * would use more and more resources. That's why we use take here
                 * instead. */
                List<Sample<NamedMessage>> samples = new ArrayList<Sample<NamedMessage>>();
                namedMessageDataReader.take(samples);
                for (Sample<NamedMessage> sample : samples) {
                    NamedMessage message = sample.getData();
                    if (message != null) {
                        if (message.userID == TERMINATION_MESSAGE) {
                            System.out.println("Termination message received: exiting...");
                            terminated = true;
                        } else {
                            System.out.println(message.userName + " : " + message.content);
                        }
                    }
                }

                /* Sleep for some amount of time, as not to consume too much CPU cycles. */
                try {
                    Thread.sleep(100);
                } catch (InterruptedException ignore) { /* Do nothing */ }
            }
        } catch (DDSException e) {
            System.out.println("Error occurred. Terminating MessageBoard (" + e.getMessage() + ")");
        }
        System.out.println("MessageBoard has terminated.");
    }
}
